kernel/posix/semaphore.c

```diff
@@ -136,3 +139,342 @@ int sem_wait(sem_t *semaphore)
+#define NSEM_SYS_OPEN_MAX  (32) /* System limit of simultaneously open nsem */
+#define NSEM_PROC_OPEN_MAX  (4) /* Process limit of simultaneously open nsem */
+
+/*
+ * Private named semaphore
+ */
+struct _nsem {
+	sem_t sem;
+	const char *name;
+	bool to_free; /* Flag set by unlink */
+	bool in_use;  /* Used to check if this spot is available */
+};
+static unsigned int sem_named_n;
+static struct _nsem sem_named[NSEM_SYS_N_MAX];
+
+/*
+ * Given a sem_t value, return the _nsem object containing it.
+ * This needs to be updated if we change the order of the elements in
+ * struct _nsem.
+ */
+#define nsem_container_of(sem_t_ptr) ((struct _nsem *)sem_t_ptr)
+
+/*
+ * Opened table entries
+ */
+struct _nsem_otable {
+	struct _nsem *nsem_ptr;
+	k_tid_t proc;
+};
+static struct _nsem_otable nsem_otable[NSEM_SYS_OPEN_MAX];
+
+/*
+ * Private functions.
+ */
+
+/*
+ * Only called if we are sure it does not exist
+ */
+static inline struct _nsem *nsem_alloc(const char *name)
+{
+	struct _nsem *it, *lmt;
+
+	it = sem_named;
+	lmt = sem_named + NSEM_SYS_N_MAX;
+
+	while (it != lmt) {
+		if (!it->in_use) {
+			it->in_use = true;
+			it->to_free = false;
+			it->name = name;
+
+			return it;
+		}
+		++it;
+	}
+
+	return NULL;
+}
+
+static inline void nsem_free(struct _nsem *nsem)
+{
+	nsem->name = NULL;
+	nsem->in_use = false;
+	nsem->to_free = false;
+	sem_destroy(&nsem->sem); /* This should never fail */
+}
+
+/*
+ * NULL if not found, pointer to the object if found.
+ * In sem_named there can be several elements with the same name, but all
+ * except one must have the field "to_free" == 0, thus, the rest are waiting to
+ * be closed.
+ */
+static inline struct _nsem *nsem_find_by_name(const char *name)
+{
+	struct _nsem *it, *lmt;
+
+	it = sem_named;
+	lmt = sem_named + NSEM_SYS_N_MAX;
+
+	while (it != lmt) {
+		if (!strncmp(name, it->name, PATH_MAX) &&
+		    it->in_use && !it->to_free)
+			return it;
+		++it;
+	}
+
+	return NULL;
+}
+
+static inline struct _nsem_otable *nsem_otable_alloc(k_tid_t proc,
+						     struct _nsem *nsem)
+{
+	struct _nsem_otable *it, *lmt;
+
+	it = nsem_otable;
+	lmt = nsem_otable + NSEM_SYS_N_MAX;
+
+	while (it != lmt) {
+		if (!it->nsem_ptr) {
+			it->nsem_ptr = nsem;
+			it->proc = proc;
+
+			return it;
+		}
+		++it;
+	}
+
+	return NULL;
+}
+
+static inline void nsem_otable_free(struct _nsem_otable *ote)
+{
+	ote->nsem_ptr = NULL; /* Flag to mark as non used */
+}
+
+/*
+ * Find in opened table an entry matching with arguments.
+ * Returns NULL if not found, or invalid arguments.
+ */
+static inline struct _nsem_otable *nsem_otable_find(k_tid_t proc,
+						    struct _nsem *nsem)
+{
+	struct _nsem_otable *it, *lmt;
+
+	if (!nsem)
+		return NULL;
+
+	it = nsem_otable;
+	lmt = nsem_otable + NSEM_SYS_OPEN_MAX;
+
+	while (it != lmt) {
+		if (it->nsem_ptr == nsem && it->proc == proc) {
+			return it;
+		}
+		++it;
+	}
+
+	return 0;
+}
+
+static inline unsigned int nsem_otable_count_refs(struct _nsem *nsem)
```

Comment:
Not understanding this.  Why not maintain a simple reference count at sem_open/sem_unlink time?  Why are you having to count the references dynamically at runtime?

Reply:
Did it for sake of consistency, having a reference counter of entries in a table is extra memory per semaphore, but I'm not opposed to the idea either.